File: src/main/java/com/sinobest/spark/apps/cores/java/WordCountLocal.java
package com.sinobest.spark.apps.cores.java;


import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.util.Arrays;

/**
 * Created by Administrator on 2016/11/17 0017.
 * 本地运行模式
 */
public class WordCountLocal {

    public static void main(String[] args){

        /**
         * 创建SparkContext  设置Spark程序运行的配置信息
         * 创建JavaSparkConf   Spark的唯一入口,实际还是调用了Scala代码 SparkContext
         */
        SparkConf conf = new SparkConf().setAppName("Spark WordCount writer by Java").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);

        /**
         * 更加SparkContext读取本地文件创建RDD
         * 并调用回调函数对lines进行单词切分
         * 数据处理流程：textFile 通过创建hadoop file 并且new了一个HadoopRDD读取本地磁盘或者集群HDFS分布式文件，
         * 并且以分片的模式，存于集群中。
         * 基于HadoopRDD产生的Partition去掉行的key,生成对应MapPartionRDD
         */
        JavaRDD<String> lines = sc.textFile("file:/E://文档//大数据//Spark//源码//spark1.6.0//spark-1.6.0-bin-hadoop2.6/README.md");
        //JavaRDD<String> lines = sc.textFile("hdfs://hadoop01.com.cn:8020/tmp/CHANGES.txt");

        /**
         * 对每个MapPartion中的每一行进行单词切分，并合并成一个大的单词实例的集合
         * flatMap产生的依旧是MapPartionRDD ，产生拆分过的集合
         * 对每个Partion的每一行进行单词切分并合并成一个大的单词实例的集合
         */
        JavaRDD<String> words = lines.flatMap((FlatMapFunction<String, String>) line -> Arrays.asList(line.split(" ")));

        /**
         * 对单词切分的基础上对每个单词计数为1
         * MapPartionRDD集合切分并且 每个单词计数为1
         * 在产生了一个MapPartionRDD (hallo 1) (spark 1) 变成kayValue的方式
         */
        JavaPairRDD<String, Integer> pairs = words.mapToPair((PairFunction<String, String, Integer>) word -> new Tuple2<>(word, 1));

        /**
         * 在每个单词的计算的基础上，对相同的单词进行累加1
         * reduceByKey
         *    1 、 shuffle之前 进行localReduce操作，主要负责本地局部统计，并把统计结果按照分区策略放到不同的file中
         *    2、
         */
        JavaPairRDD<String, Integer> wordCount = pairs.reduceByKey((Function2<Integer, Integer, Integer>) (v1, v2) -> v1 + v2);


        /**
         * 以上的全部步骤，属于一个Stage, Stage的操作是基于内存迭代的
         */

        /**
         * 打印统计的内容
         */
        wordCount.foreach((VoidFunction<Tuple2<String, Integer>>) pairs1 -> System.out.println(pairs1._1 + ":" + pairs1._2));

    }
}

File: src/main/code/com/bigdata/spark/apps/cores/WordCountLocal.scala
package com.bigdata.spark.apps.cores

import com.bigdata.spark.apps.submit.SparkSubmitUtil
import org.apache.spark.{SparkConf, SparkContext}

/**
  * Created by liulv on 2016/11/13 0013.
  *
  * Spark Word Count 本地模式
  */
object WordCountLocal {

  def main(args: Array[String]): Unit = {

    //创建SparkConf
    //设置设置本地执行，不需要安装spark集群
    //比如setMaster 设置集群Master ULR,如果设置local是Spark在本地运行
    val conf = new SparkConf()
    conf.setAppName("Wow, My first Spark APP!")
    //本地模式
    conf.setMaster("local")

    SparkSubmitUtil.setJobHistory(conf)

    //创建SparkContext对象
    //SparkContext是Spark唯一的入口程序，无论是采用Scala,Java
    //SParkContext的核心作业：初始化Spark运用程序所需要的核心组件
    //同时还会负责SPark程序往Master注册程序等，SParkContext是Spark中最为至关重要的一个对象
    //创建SparkContext通过传递SparkConf实例来定制Spark运行的具体的参数和对象
    val sc = new SparkContext(conf)

    //根据具体的来源（HDFS, HBase, Loacl, Fs, DB ,S3等）通过SparkContext来创建RDD
    //RDD的基本有三种方式：1 根据外部的数据来源（HDFS）2 根据Scala集合 3 由其他RDD来创建
    //数据会根据RDD划分为一系列的Partitions,分配到每个的Partition的数据属于一个Task的处理范畴
    // 本地模式
    val lines = sc.textFile("file:/G://文档//Spark//源码//spark1.6.0//spark-1.6.0-bin-hadoop2.6//README.md", 1) //读取本地文件，并设置为Partition

    //对初始的RDD
    //第一步拆分每个单词
    val words = lines.flatMap { line => line.split(" ") }

    //在每个单词计算1的基础上根据出现次数进行累加
    //MapPartionRDD
    val pairs = words.map { word => (word, 1) }

    /**
      *  广告点击排名，通过倒序排序说明：
      *
      *  1、reduceByKey(_ + _) 对每个出现相同的key,进行累加组合为新的组合
      *  2、map(pairs => (pairs._2, pairs._1)通过key和value互相换位置，好进行相关的key进行排序
      *  3、sortByKey(false) 对map中的key进行排序，false为倒序排序，默认为正序
      *  4、最后在一次对 key 和 value进行互换位置，2次互换为最开始的key value格式
      *
      *  数据流程：
      *  1、Shuffed前
      *       shuffle之前 进行localReduce操作，主要负责本地局部统计，并把统计结果按照分区策略放到不同的file中
      *  2、产生了ShuffedRDD
      *       把上一个结果的几个类集合的分片的相同的key的进行shuffed洗牌累加
      *  3、MapPartitionRDD
      */
    val wordCounts = pairs.reduceByKey(_ + _)

    /**
      * 排序
      */
    val wordCountsSort = wordCounts.map(pairs => (pairs._2, pairs._1)).sortByKey(ascending = false).map(pairs => (pairs._2, pairs._1))

    wordCountsSort.collect.foreach(wordNumberPar => println(wordNumberPar._1 + " : " + wordNumberPar._2))

    sc.stop()

  }

}
